kong/runloop/wasm.lua

```diff
@@ -1,141 +1,537 @@
   if not ok then
-    return nil, "initial loading of proxy wasm modules failed: " .. tostring(err)
+    return nil, "failed loading filters: " .. tostring(err)
   end
 
-  return true
+  return c_plan
 end
 
 
-function _M.attach_filter_chains(ctx)
-  if not enabled then
-    return
+-- Helper method for retrieving a filter chain reference from
+-- the state table.
+--
+---@param state       kong.runloop.wasm.state
+---@param typ         integer
+---@param service_id? string
+---@param route_id?   string
+---
+---@return kong.runloop.wasm.filter_chain_reference? ref
+local function get_chain_ref(state, typ, service_id, route_id)
+  local ref
+
+  if typ == TYPE_SERVICE and service_id then
+    ref = state.by_service[service_id]
+
+  elseif typ == TYPE_ROUTE and route_id then
+    ref = state.by_route[route_id]
+
+  elseif typ == TYPE_COMBINED and service_id and route_id then
+    local routes = state.combined[service_id]
+    ref = routes and routes[route_id]
   end
 
-  local service = ctx.service or NONE
-  local route = ctx.route or NONE
+  return ref
+end
+
 
-  local filters = {}
+---
+-- Helper method for storing a new filter chain reference within
+-- the state table.
+--
+---@param state       kong.runloop.wasm.state
+---@param ref         kong.runloop.wasm.filter_chain_reference
+---@param typ         integer
+---@param service_id? string
+---@param route_id?   string
+local function store_chain_ref(state, ref, typ, service_id, route_id)
+  if typ == TYPE_SERVICE and service_id then
+    state.by_service[service_id] = ref
+
+  elseif typ == TYPE_ROUTE and route_id then
+    state.by_route[route_id] = ref
+
+  elseif typ == TYPE_COMBINED and service_id and route_id then
+    local routes = state.combined[service_id]
+
+    if not routes then
+      routes = {}
+      state.combined[service_id] = routes
+    end
 
-  for chain, err in kong.db.wasm_filter_chains:each() do
+    routes[route_id] = ref
+  end
+end
+
+
+---
+-- Build a combined filter list from 1-2 filter chain entities.
+--
+-- Disabled filter chains are skipped, and disabled filters are
+-- skipped.
+--
+-- Returns `nil` if no enabled filters are found.
+--
+---@param service_chain? kong.db.schema.entities.wasm_filter_chain
+---@param route_chain?   kong.db.schema.entities.wasm_filter_chain
+---
+---@return kong.db.schema.entities.wasm_filter[]?
+local function build_filter_list(service_chain, route_chain)
+  ---@type kong.db.schema.entities.wasm_filter[]|nil
+  local combined
+  local n = 0
+
+  if service_chain and service_chain.enabled then
+    for _, filter in ipairs(service_chain.filters) do
+      if filter.enabled then
+        n = n + 1
+        combined = combined or {}
+        combined[n] = filter
+      end
+    end
+  end
+
+  if route_chain and route_chain.enabled then
+    for _, filter in ipairs(route_chain.filters) do
+      if filter.enabled then
+        n = n + 1
+        combined = combined or {}
+        combined[n] = filter
+      end
+    end
+  end
```

Comment:
Interesting. So service-level filters run before route-level filters?
That seems backward to me because there is no way for the route level filter to run first and then put something in the request context to "hint" the service level filter to not run or provide some additional metadata. This comes to my mind because route-level plugins take precedence over service-level plugins.

What do you think?

Reply:
My thought process here was that the route filter should run last as sort of a tie-breaker.

Example: if I have a `set-response-header` filter that is attached to both a route and service and sets a `X-Foo` header, then the route filter will execute last, overwriting the header that was set by the service filter and thereby having more control over the final state of things.

There are plenty of counter examples to this though; this might be worth another group discussion.

Reply:
Yes, this requires some discussion since it has product implications and as this conversation highlights, an argument can be made either way.